2024 Importer

awisema opened the pull request.

awisema committed:
WIP: Importer models

awisema committed:
WIP: Move to use ExternalFileUtils

awisema committed:
WIP: 2024 importer changes

awisema committed:
Update CSVVersion string

awisema committed:
WIP: End-to-end import

awisema committed:
Participation table updates

awisema committed:
Update validations

awisema committed:
Merge branch 'release-86' into aw/2024-importer-185275581

# Conflicts:
#	db/warehouse_structure.sql

awisema committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

# Conflicts:
#	db/warehouse_structure.sql

eanders committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

eanders committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

eanders commented on drivers/hmis_csv_twenty_twenty_four/app/models/hmis_csv_twenty_twenty_four/importer/import_concern.rb:
```diff
@@ -0,0 +1,297 @@
+    end
+
+    # HMIS CSV FORMAT SPECIFICATIONS says under "Data Types"
+    # https://hudhdx.info/Resources/Vendors/HMIS%20CSV%20Specifications%20FY2020%20v1.8.pdf
```
This seems a bit outdated now.  I assume the information hasn't changed, maybe we should just remove the link?

eanders commented on drivers/hmis_csv_twenty_twenty_four/extensions/grda_warehouse/hud/project_extension.rb:
```diff
@@ -0,0 +1,52 @@
+      has_many :imported_items_2024, class_name: '::HmisCsvTwentyTwentyFour::Importer::Project', primary_key: [:ProjectID, :data_source_id], foreign_key: [:ProjectID, :data_source_id]
+      has_many :loaded_items_2024, class_name: '::HmisCsvTwentyTwentyFour::Loader::Project', primary_key: [:ProjectID, :data_source_id], foreign_key: [:ProjectID, :data_source_id]
+
+      def convert_to_aggregated!
+        existing = HmisCsvImporter::Aggregated::Enrollment.where(data_source_id: data_source_id, ProjectID: self.ProjectID).exists?
+        raise 'This can only be run once per data source!' if existing
+
+        data_source.update(
+          import_aggregators: {
+            'Enrollment' => ['HmisCsvImporter::Aggregated::CombineEnrollments'],
+          },
+        )
+        update(combine_enrollments: true)
+
+        copy_data_to_aggregates(enrollments, HmisCsvImporter::Aggregated::Enrollment)
+        copy_data_to_aggregates(exits, HmisCsvImporter::Aggregated::Exit)
+      end
+
+      private def copy_data_to_aggregates(scope, destination_class)
+        pseudo_importer_log = HmisCsvImporter::Importer::ImporterLog.new(id: 0)
+        scope.find_in_batches do |batch|
+          look_asides = []
+          batch.each do |row|
+            data = row.slice(row.class.hmis_structure(version: '2020').keys)
+            data.merge!(
+              source_type: row.class.name,
+              source_id: row.id,
+              data_source_id: data_source_id,
+              importer_log_id: pseudo_importer_log.id,
+              pre_processed_at: Time.current,
+            )
+            new_item = destination_class.new(data)
+            new_item.set_source_hash
+            look_asides << new_item
+          end
+          destination_class.import(look_asides)
+        end
+      end
+    end
```
Are these included in the 2022 extension as well?  Should we version them, or maybe remove the old ones, or move them to some shared concern?  Seems odd to inject the same thing multiple times.

awisema replied:
Yes -- this was specifically for the 2020 -> 2022 migration with clients that were using the versioned aggregation tables, which we removed. Deleted.

eanders commented on drivers/hmis_csv_twenty_twenty_four/extensions/grda_warehouse/hud/project_extension.rb:
```diff
@@ -0,0 +1,52 @@
+        scope.find_in_batches do |batch|
+          look_asides = []
+          batch.each do |row|
+            data = row.slice(row.class.hmis_structure(version: '2020').keys)
```
Does this need updating?
```suggestion
            data = row.slice(row.class.hmis_structure(version: '2024').keys)
```

awisema committed:
Clean up from review.

eanders reviewed (approved):
@awisema any chance we could add  a smoke test?  It'd be great to take a 2024 file set and confirm that we have the right number of warehouse records after import.

Otherwise, I think this looks great.  Thank you.

awisema committed:
Set timestamps on HMISParticipation from Project
Was setting the timestamps to reflect when the data was created. For exports, this breaks the effective end date.

awisema committed:
Disable 2022 tests

awisema committed:
2024 smoke test

awisema committed:
Commit disabled 2022 importer tests for reference

awisema committed:
Adjust delete empty enrollments to new NBN project type

awisema committed:
Adjust EnforceRelationshipToHoh to use new gender column name

awisema committed:
Remove HMISParticipatingProject from Project validations

awisema committed:
Move green importer tests

eanders commented on db/reporting_structure.sql:
Pretty sure this shouldn't have been included.

eanders committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

awisema committed:
Fix enrollment validations test count

awisema committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

awisema committed:
Don't step on the TRANSFORMS constant
Running the transformer more than once was failing because the references were getting transformed in place.

eanders committed:
Merge branch 'ea/fy-2024' into aw/2024-importer-185275581

awisema committed:
Disable PATH tests that 2024 breaks

awisema committed:
Disable SPM tests that 2024 breaks

awisema committed:
Fix some tests that were assuming all ES were project_type 1
